src/main/scala/cromwell/Main.scala

```diff
@@ -21,167 +20,184 @@ object Actions extends Enumeration {
 }
 
 object Main extends App {
-  private def initLogging() = {
-    val LogRootDefault = "."
-    val LogModeDefault = "CONSOLE"
-    val LogLevelDefault = "INFO"
-
-    val logRoot = Option(System.getProperty("LOG_ROOT")) match {
-      case None =>
-        System.setProperty("LOG_ROOT", LogRootDefault)
-        LogRootDefault
-      case Some(x) => x
-    }
-
-    if (Option(System.getProperty("LOG_MODE")).isEmpty)
-      System.setProperty("LOG_MODE", LogModeDefault)
-
-    if (Option(System.getProperty("LOG_LEVEL")).isEmpty)
-      System.setProperty("LOG_LEVEL", LogLevelDefault)
-
-    try {
-      Files.createDirectories(Paths.get(logRoot))
-    } catch {
-      case e: FileAlreadyExistsException =>
-      case e: Throwable =>
-        System.err.println(s"Could not create log directory: $logRoot")
-        e.printStackTrace()
-        System.exit(1)
-    }
-  }
+  // scala.App's DelayedInit is tricky, as the docs say. Split the main logic out so that it's easier to test.
+  new Main(args)
+}
 
-  initLogging()
+class Main private[cromwell](enableSysExit: Boolean, managerSystem: () => WorkflowManagerSystem) {
```

Comment:
One cannot easily modify variables on an object that implements `App`/`DelayedInit`, because the scala compiler does all sorts of magic to the class, and your property changes will be erased. If one wants to unit test individual methods like `run`, you run into strange side effects, for example, trying to externally set a `WorkflowActorSystem` would just end up with `null` when you ran later ran `Main.run()`:

- [Weird initialisation when override val met DelayedInit in Scala](http://stackoverflow.com/questions/17716975/weird-initialisation-when-override-val-met-delayedinit-in-scala)
- [what happens in scala when loading objects that extends App?](http://stackoverflow.com/questions/13009265/what-happens-in-scala-when-loading-objects-that-extends-app)
- [So what can you use delayedInit for safely?](http://stackoverflow.com/questions/18582709/so-what-can-you-use-delayedinit-for-safely)
- [In Scala; should I use the App trait?](http://stackoverflow.com/questions/24437423/in-scala-should-i-use-the-app-trait)
- etc. etc. etc.

I eventually left the scala-esque App trait there instead of switching to Java-hybrid `def main(args: Array[String]): Unit`, but left the functionality in an class that could actually be unit tested without `DelayedInit` side effects.